src/addons/meta/runtime.c

```diff
@@ -0,0 +1,328 @@
+  if (!ctx)
+    return;
+
+  ecs_meta_rtt_ctx_t *lifecycle_ctx = (ecs_meta_rtt_ctx_t *) ctx;
+
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->ctors, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->dtors, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->moves, ecs_meta_rtt_call_data_t);
+  ecs_vec_fini_t(NULL, &lifecycle_ctx->copys, ecs_meta_rtt_call_data_t);
+
+  ecs_os_free(ctx);
+}
+
+static ecs_meta_rtt_ctx_t *flecs_meta_rtt_configure_hooks(ecs_world_t *world, const ecs_type_info_t *ti, bool ctor,
+                                                          bool dtor, bool move, bool copy) {
+  ecs_type_hooks_t hooks = ti->hooks;
+  flecs_meta_rtt_free_lifecycle_ctx(hooks.lifecycle_ctx);
+  ecs_meta_rtt_ctx_t *rtt_ctx = NULL;
+  if (ctor || dtor || move || copy) {
+    rtt_ctx = ecs_os_malloc_t(ecs_meta_rtt_ctx_t);
+    ecs_vec_init_t(NULL, &rtt_ctx->ctors, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->dtors, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->moves, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->copys, ecs_meta_rtt_call_data_t, 0);
+    hooks.lifecycle_ctx = rtt_ctx;
+    hooks.lifecycle_ctx_free = flecs_meta_rtt_free_lifecycle_ctx;
+
+    if (ctor) {
+      hooks.ctor = flecs_meta_rtt_ctor;
+    }
+    if (dtor) {
+      hooks.dtor = flecs_meta_rtt_dtor;
+    }
+    if (move) {
+      hooks.move = flecs_meta_rtt_move;
+    }
+    if (copy) {
+      hooks.copy = flecs_meta_rtt_copy;
+    }
+  } else {
+    hooks.lifecycle_ctx = NULL;
+  }
+  ecs_set_hooks_id(world, ti->component, &hooks);
+  return rtt_ctx;
+}
+
+// Checks if a struct member's types have hooks installed. If so, it generates and installs required hooks for the
+// struct type itself. These hooks will invoke the member hooks when necessary
+// returns true if hooks were actually configured, false otherwise (struct did not need hooks)
+static void flecs_meta_rtt_init_default_hooks_struct(ecs_world_t *world, ecs_entity_t component,
+                                                     const ecs_type_info_t *ti) {
+  // Obtain struct information to figure out what members it contains:
+  const EcsStruct *struct_info = ecs_get(world, component, EcsStruct);
+  ecs_assert(struct_info != NULL, ECS_INTERNAL_ERROR, NULL);
+
+  // These flags will be set to true if we determine we need to generate a hook of a particular type:
+  bool ctor_hook_required = false;
+  bool dtor_hook_required = false;
+  bool move_hook_required = false;
+  bool copy_hook_required = false;
+
+  // Iterate all struct members and see if hooks are required:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
+    ecs_member_t *m = ecs_vec_get_t(&struct_info->members, ecs_member_t, i);
+    const ecs_type_info_t *member_ti = ecs_get_type_info(world, m->type);
+    ctor_hook_required |= member_ti->hooks.ctor && member_ti->hooks.ctor != flecs_default_ctor;
```

Comment:
I think so, because I don't know who set that `flecs_default_ctor` constructor, perhaps even the user did in a custom component.

Reply:
If the user explicitly provided `flecs_default_ctor` it should not be overridden. This should be no different than a user providing a custom function that initializes to 0. If possible I'd prefer it if there are no special semantics for `flecs_default_ctor`.

Reply:
It is possible, but I'd lose the possibility to merge contiguous `flecs_default_ctor` in a struct. It is a great flag indicating the user just wants a memset(0). Since it is public, `FLECS_API`, we could document that the call may be merged together with others, but honestly I think this change is transparent to the user.

A struct like this:
```c
struct Person {
    char* name; // ecs_string_t
    char* last_name; // ecs_string_t
    char* city; // ecs_string_t
};
```
has 3 fields, each would call `flecs_default_ctor` for `ecs_string_t`. The proposed change detects this is the case and sets `flecs_default_ctor` at the struct level, instead of a call list with 3 items pointing to independent `flecs_default_ctor`s.

Additionally, I cannot distinguish between the author of a type not defining a ctor or explicitly setting `flecs_default_ctor`. By the time my code runs, I've lost that distinction; I only see `flecs_default_ctor`.

If the user wants to opt out of `flecs_default_ctor` merging, they can make their own ctor, but what would be the point?

Reply:
It's more a principle of good API design, it should not override things that a user sets explicitly. If I do this:
```cpp
ecs_set_hooks(world, T, {
  .ctor = flecs_default_ctor
});
```
... then I expect that unconditionally, this always holds true:
```cpp
assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor);
```
It's true that this means you can't "collapse" multiple `flecs_default_ctor`'s, but that's a performance optimization that can be addressed in a different way in the future that doesn't have observable side effects for the user.

Reply:
> `assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor);`

This holds true. I am not replacing your hook, because `T` is a component, meaning `EcsType.existing = true` and the new code does not run. The full example would be:

```c
  struct T {
    int value;
  };
  ECS_COMPONENT(world, T);
  ecs_set_hooks(world, T, {.ctor = flecs_default_ctor});
```

Another case would be that `T` is a field of a rtt struct. In that case, I just "consider" that you set T's ctor to `flecs_default_ctor` and will (in the future) create a collapsed call data, but I am not modifying the original hook. The optimization already in place is that if all fields in a struct happen to have `flecs_default_ctor` or `NULL`, then the resulting rtt struct's ctor will be `flecs_default_ctor` directly and we won't have a call per field and we won't have a `lifecycle_ctx` at all. That is actually the behavior prior to this PR.

```c
  struct T {
    int value;
  };
  ecs_entity_t t = ECS_COMPONENT(world, T);
  ecs_set_hooks(world, T, {.ctor = flecs_default_ctor});

  ecs_entity_t rtt_struct = ecs_struct(world, {.entity = ecs_entity(world, {.name = "RttStruct"}),
                                   .members = {{.name = "x", .type = t},
                                               {.name = "y", .type = ecs_id(ecs_u32_t)}});

  assert(ecs_get_type_info(world, T)->hooks.ctor == flecs_default_ctor); // holds true
  assert(ecs_get_type_info(world, rtt_struct)->hooks.ctor == flecs_default_ctor); // holds true

```